File: back-end/src/main/java/ccu/four/backend/models/Sensor.java
package ccu.four.backend.models;

import ccu.four.backend.enums.Measures;
import ccu.four.backend.models.note.SensorNote;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import org.jetbrains.annotations.NotNull;

import javax.persistence.*;
import java.util.*;

@Entity
@NamedNativeQueries({
        @NamedNativeQuery(
                name = "sensor_values",
                query = "SELECT DISTINCT v.time as date, v.value FROM sensor s JOIN sensor_values v ON v.sensor_identifier = ?1 ORDER BY v.time DESC LIMIT ?2"
        ),
        @NamedNativeQuery(
                name = "sensor_targets",
                query = "SELECT DISTINCT v.time as date, v.value FROM sensor s JOIN sensor_target_values v ON v.sensor_identifier = ?1 ORDER BY v.time DESC LIMIT ?2"
        ),
        @NamedNativeQuery(
                name = "add_sensor_value",
                query = "INSERT INTO sensor_values (sensor_identifier, value, time) VALUES (?1, ?2, ?3)"
        ),
        @NamedNativeQuery(
                name = "add_sensor_target",
                query = "INSERT INTO sensor_target_values (sensor_identifier, value, time) VALUES (?1, ?2, ?3)"
        ),
        @NamedNativeQuery(
                name = "getInitialSensors",
                query = "SELECT * FROM sensor where team_id = ?"
        )
})
public class Sensor {

    @Id
    private String identifier;
    private int idx;
    private String sensorType;
    private double minValue;
    private double maxValue;
    private Measures unit;
    private String readableName;
    private String description;
    private Double step;

    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "sensor_values")    // Table and foreign key name
    @MapKeyColumn(name = "time")                // Column name for the Date type.
    @Column(name = "value")                     // Column name for the Double type.
    @JsonIgnore
    private Map<Date, String> values = new LinkedHashMap<>();

    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "sensor_target_values")  // Table and foreign key name
    @MapKeyColumn(name = "time")                    // Column name for the Date type.
    @Column(name = "value")                         // Column name for the Double type.
    @JsonIgnore
    private Map<Date, String> targetValues = new LinkedHashMap<>();

    @JsonBackReference
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "teamId")
    private Team team;

    @JsonManagedReference
    @OneToMany(fetch = FetchType.LAZY, mappedBy = "sensor")

    private List<SensorNote> notes;

    public Sensor() {
        this.identifier = buildId("null", 0);
    }

    public Sensor(@NotNull Sensor sensor) {
        this(sensor.idx, sensor.sensorType, sensor.minValue, sensor.maxValue, sensor.unit, sensor.readableName, sensor.description, new HashMap<>(sensor.values), new HashMap<>(sensor.targetValues), sensor.step, null);
    }

    public Sensor(int idx, String sensorType, double minValue, double maxValue, Measures unit, String readableName, String description, Map<Date, String> values, Map<Date, String> targetValues, Double step, Team team) {
        this.idx = idx;
        this.sensorType = sensorType;
        this.minValue = minValue;
        this.maxValue = maxValue;
        this.unit = unit;
        this.readableName = readableName;
        this.description = description;
        this.values = values;
        this.targetValues = targetValues;
        this.team = team;
        this.step = step;
    }

    public Sensor(int idx, String sensorType, double minValue, double maxValue, Measures unit, String readableName, String description, Double step, Team team) {
        this(idx, sensorType, minValue, maxValue, unit, readableName, description, new HashMap<>(), new HashMap<>(), step, team);
    }

    public Sensor(String sensorType, Team team) {
        this.identifier = buildId(sensorType, team);
        this.sensorType = sensorType;
        this.team = team;
    }

    public static String buildId(String sensorType, long teamId) {
        return sensorType + teamId;
    }

    public static String buildId(String sensorType, Team team) {
        return buildId(sensorType, team.getId());
    }

    public void addTarget(Map.Entry<Date, String> entry) {
        addTarget(entry.getKey(), entry.getValue());
    }

    public void addValue(Map.Entry<Date, String> entry) {
        addValue(entry.getKey(), entry.getValue());
    }

    public void addTarget(Date date, String value) {
        targetValues.put(date, value);
    }

    public void addValue(Date date, String value) {
        values.put(date, value);
    }

    @JsonIgnore
    public Map.Entry<Date, String> getLatestValue() {
        return values.entrySet().stream()
                .max(Map.Entry.comparingByKey())
                .orElse(null);
    }

    @JsonIgnore
    public Map.Entry<Date, String> getLatestTarget() {
        return targetValues.entrySet().stream()
                .max(Map.Entry.comparingByKey())
                .orElse(null);
    }

    public void setIdx(int idx) {
        this.idx = idx;
    }

    public void setSensorType(String sensorType) {
        this.identifier = buildId(sensorType, team);
        this.sensorType = sensorType;
    }

    public void setMinValue(double minValue) {
        this.minValue = minValue;
    }

    public void setMaxValue(double maxValue) {
        this.maxValue = maxValue;
    }

    public void setUnit(Measures unit) {
        this.unit = unit;
    }

    public void setReadableName(String readableName) {
        this.readableName = readableName;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getIdx() {
        return idx;
    }

    public String getSensorType() {
        return sensorType;
    }

    public double getMinValue() {
        return minValue;
    }

    public double getMaxValue() {
        return maxValue;
    }

    public Measures getUnit() {
        return unit;
    }

    public String getReadableName() {
        return readableName;
    }

    public String getDescription() {
        return description;
    }

    public List<SensorNote> getNotes() {
        return notes;
    }

    public Team getTeam() {
        return team;
    }

    public void setTeam(Team team) {
        this.identifier = buildId(this.sensorType, team);
        this.team = team;
    }

    public Map<Date, String> getValues() {
        return values;
    }

    public Map<Date, String> getTargetValues() {
        return targetValues;
    }

    public String getIdentifier() {
        return identifier;
    }
}

File: front-end/src/app/models/sensor.ts
import {Note} from "./notes/note";
import {SensorNote} from "./notes/sensorNote";
import {Measures} from "./measures";

export class Sensor {

  private _identifier: string;
  private _idx: number;
  private _sensorType: string;
  private _minValue: number;
  private _maxValue: number;
  private _unit: Measures;
  private _readableName: string;
  private _description: string;
  private _notes: SensorNote[];
  private _valueStep: number;
  private _targets: [{ "date": Date, "value": string }]
  private _values: [{ "date": Date, "value": string }]

  constructor(idx?: number, sensorType?: string, minValue?: number, maxValue?: number, unit?: Measures, readableName?: string, description?: string, valueStep?: number, identifier?: string) {
    this._identifier = identifier;
    this._idx = idx;
    this._sensorType = sensorType;
    this._minValue = minValue;
    this._maxValue = maxValue;
    this._unit = unit;
    this._readableName = readableName;
    this._description = description;
    this._notes = [];
    this._valueStep = valueStep;
  }

  public addNote(note: SensorNote): void {
    this._notes.push(note);
  }

  public removeNote(note: Note): boolean {
    for (let i = 0; i < this._notes.length; i++) {
      if (this._notes[i] === note) {
        this._notes.splice(i, 1)
        return true;
      }
    }
    throw new Error('Note not found');
  }

  get notes(): SensorNote[] {
    return this._notes;
  }

  set minValue(value: number) {
    this._minValue = value;
  }

  set maxValue(value: number) {
    this._maxValue = value;
  }

  set unit(value: Measures) {
    this._unit = value;
  }

  set readableName(value: string) {
    this._readableName = value;
  }

  set description(value: string) {
    this._description = value;
  }

  get idx(): number {
    return this._idx;
  }

  get sensorType(): string {
    return this._sensorType;
  }

  get minValue(): number {
    return this._minValue;
  }

  get maxValue(): number {
    return this._maxValue;
  }

  get unit(): Measures {
    return this._unit;
  }

  get readableName(): string {
    return this._readableName;
  }

  get description(): string {
    return this._description;
  }


  get valueStep(): number {
    return this._valueStep;
  }


  get identifier(): string {
    return this._identifier;
  }

  set identifier(value: string) {
    this._identifier = value;
  }

  set idx(value: number) {
    this._idx = value;
  }

  set sensorType(value: string) {
    this._sensorType = value;
  }

  set notes(value: SensorNote[]) {
    this._notes = value;
  }

  set valueStep(value: number) {
    this._valueStep = value;
  }

  set targets(value: [{ date: Date; value: string }]) {
    this._targets = value;
  }

  set values(value: [{ date: Date; value: string }]) {
    this._values = value;
  }
}
